README.md
# Tiny Browser Framework

[![npm](https://img.shields.io/npm/v/tiny-browser-framework.svg)]()
[![Build Status](https://travis-ci.org/thedumbterminal/TinyBrowserFramework.svg?branch=master)](https://travis-ci.org/thedumbterminal/TinyBrowserFramework)

Tiny browser web framework, which communicates to the server via a websocket. All DOM changes are performed using HTML provided by the server.

Minified size is **2.23** Kb which when transmitted compressed is less than 1Kb!

## Usage

Add the following script tag at the bottom of your HTML as follows:

    <script src="https://unpkg.com/tiny-browser-framework/dist/index.min.js"></script>

Then add `data-url` attributes to any buttons on your pages, the value of this would be the data that will be sent over the websocket to the server to process.

    <button data-url="/something?a=b">An example button</button>

No change is required for any forms present on your pages.

Full example HTML available at [example/example.html](example/example.html).

## Specification

The server does all the heavy lifting and the browser is just used for updating fragments of HTML on the page, which keeps the clientside logic minimal.

Any state, authentication or persistence is not a concern of this client framework, these can also be managed by the server, but might involve a client side element such as setting cookies or using other JS libs for OAuth etc.

Data is sent to the server when an augmented element on the page is triggered, the server may then response with zero or more fragments of HTML to update the page.

The server may also return HTML fragments to the client based on server events rather than user actions, possible scenarios for this could be:

* Message received from pub/sub topic.
* Time based logic.
* Server state change, such as a multi-user system.
* Long running job finishes, such as encoding an uploaded video.

This means that the browser does not need to wait for the server to return a final response. It could return a number of fragments in a timely manner to provide a helpful UI until a long operation finishes.

### Client

Forms are augmented automatically using their action URLs.

Other clickable elements will be augmented if the `data-url` property is set. The value is the server URL where a GET request to handle the action resides.

### Server

A server must return JSON from websocket requests to `/websocket` in the following format:

    [{
    	"action": "append|replace",
    	"container": "ID of container element to update",
    	"content": "HTML string"
    }]

As this is an array the server may return 0 or more elements that need to updated at once.

An example of this is:

    [{
    	"action": "append",
    	"container": "todos",
    	"content": "<li>Another todo</li>"
    }, {
    	"action": "replace",
    	"container": "numTodos",
    	"content": "You have <b>5</b> todos."
    }]

The above example would add another child to the element with the ID of "todos" and replace the contents of the element with the ID of "numTodos".

## Live demo

[https://tiny-browser-framework.herokuapp.com/](https://tiny-browser-framework.herokuapp.com/)

This is a simple todo list app, which also supports a reminder for an item.

This example shows the following:

* Using a cookie session with a websocket.
* Direct DOM change from user action.
* Indirect DOM change from user action.
* DOM change via timed event.

## Developing

Install the package:

    npm install tiny-browser-framework

Then create a script tag in your HTML to reference the `src/index.js` file.

See the [example directory](example/) for more information.

### Included node.js express server example

To run the example server use:

    nvm use
    npm install
    npm start

Then open [http://localhost:3000/](http://localhost:3000/) in your browser.

## Compatibility

Browser shims for modern browser functionality are not included in this project.

The following would be required:

* [Array.prototype.forEach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach)
* [Websockets](https://github.com/Modernizr/Modernizr/wiki/HTML5-Cross-Browser-Polyfills#web-sockets)

## Todo

* Implement more actionable elements, such as images or links etc.
* Implement more ways to trigger an element such as `onblur`, `onfocus`, `onscroll` etc.



CHANGELOG.md
# Changelog

## v1.1.1
(20/02/2018)

* Updating dependencies.
* Improved JS minification.

## v1.1.0
(16/07/2017)

* Will now reconnect when the websocket is closed.

## v1.0.4
(07/03/2017)

* Secure websockets now supported.

## v1.0.3
(07/03/2017)

* Websocket connection will now use correct hostname.

## v1.0.2
(07/03/2017)

* Added dist directory to npm release.

## v1.0.1
(05/03/2017)

* Documentation updated.

## v1.0.0
(05/03/2017)

* Initial release


src/index.js
'use strict';

var TBF = function(){
	this._setupListeners();
	this._setupWebsocket();
};

TBF.prototype._setupWebsocket = function(){
	var socketUrl = window.location.protocol.replace('http', 'ws') + '//' + window.location.host + '/websocket';
	this._websocket = new WebSocket(socketUrl);
	var self = this;
	this._websocket.onclose = function(event){
		setTimeout(function(){
			self._websocket.readyState > 1 && self._setupWebsocket();
		}, 1000);
	}
	this._websocket.onmessage = function(event){
		var jsons = JSON.parse(event.data);
		jsons.forEach(function(json){
			self._handleResponse(json);
		});
	};
};

TBF.prototype._augmentInterface = function(){
	var tags = document.getElementsByTagName('BUTTON');
	for(var i = 0; i < tags.length; i++){
		this._augmentButton(tags.item(i));
	}
	tags = document.getElementsByTagName('FORM');
	for(var i = 0; i < tags.length; i++){
		this._augmentForm(tags.item(i));
	}
};

TBF.prototype._augmentButton = function(ele){
	if(!ele.dataset.url){
		return;
	}
	if(ele.onclick){	// already done
		return;
	}
	console.log('augmenting button: ', ele);
	var self = this;
	ele.onclick = function(){
		self._elementActivated(ele);
	}
};

TBF.prototype._augmentForm = function(ele){
	if(ele.onsubmit){	// already done
		return;
	}
	console.log('augmenting form: ', ele);
	var self = this;
	ele.onsubmit = function(){
		self._elementActivated(ele);
		return false;
	}
};

TBF.prototype._elementActivated = function(ele){
	console.log('element activated:', ele);
	this._websocket.send(this._getActionURL(ele));
};

TBF.prototype._getActionURL = function(ele){
	if(ele.tagName === 'BUTTON'){
		return ele.dataset.url;
	}
	else if(ele.tagName === 'FORM'){
		var kvPairs = [];
		for(var i = 0; i < ele.elements.length; i++){
			var e = ele.elements[i];
			kvPairs.push(encodeURIComponent(e.name) + '=' + encodeURIComponent(e.value));
		}
		var queryString = kvPairs.join('&');
		return ele.action + '?' + queryString;
	}
	throw new Error('Unknown element type');
};

TBF.prototype._handleResponse = function(json){
	console.log('got json:', json);
	var container = document.getElementById(json.container);
	if(!container){
		throw new Error('container not found: ' + json.container);
	}
	if(json.action === 'replace'){
		container.innerHTML = json.content;
	}
	else if(json.action === 'append'){
		var template = document.createElement('template');
		template.innerHTML = json.content;
		for(var i = 0; i < template.content.childNodes.length; i++){
			container.appendChild(template.content.childNodes.item(i));
		}
	}
	else{
		throw new Error('Unknown action: ', json.action);
	}
};

TBF.prototype._setupListeners = function(){
	var self = this;
	document.addEventListener('DOMContentLoaded', function() {
		self._augmentInterface();
	});
	document.addEventListener('DOMSubtreeModified', function() {
		self._augmentInterface();
	});
	window.onbeforeunload = function(){
		self._websocket && self._websocket.close();
	};
};

window.tbfInstance = new TBF();


example/server.js
'use strict'
//Server for demo application

const session = require('express-session')
const express = require('express')
const url = require('url')
const qs = require('qs')
const http = require('http')
const WebSocket = require('ws')
const compression = require('compression')
const info = require('../package.json')

const app = express()

const sessionHandler = session({
	resave: true,
	saveUninitialized: true,
	secret: 'really secure'
})

app.use(compression())
app.use(sessionHandler)
app.use(express.static('src'))

const server = http.createServer(app)

const wss = new WebSocket.Server({
	server: server,
	path: '/websocket'
})

const port = process.env.PORT || 3000
server.listen(port, () => {
	console.log('Listening on %d', server.address().port)
});

//Return the starting HTML to the client
//Just inlined it here as this is just an example server.
app.get('/', (req, res, next) => {
	const scriptSrc = process.env.NODE_ENV === 'production' ? `https://unpkg.com/tiny-browser-framework@${info.version}/dist/index.min.js` : 'index.js'
	const response = `
	<!DOCTYPE html>
	<html lang="en">
		<head>
			<meta charset="utf-8"/>
			<meta http-equiv="X-UA-Compatible" content="IE=edge">
			<meta name="viewport" content="width=device-width, initial-scale=1">
			<title>Tiny Browser Framework: Todo Example</title>
			<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
		</head>
		<body>
			<div class="container">
				<div class="page-header">
					<h1>Todo Example</h1>
				</div>
				<div id="intro" data-url="/intro"></div>
				<div id="count">${renderTodoCount(req.session)}</div>
				<ul id="lists" class="list-group">
					${renderTodos(req.session)}
				</ul>
				<form action="add">
					<div class="form-group">
						<label>Description</label><input type="text" name="description" />
					</div>
					<div class="form-group">
						<label>Reminder (seconds)</label><input type="range" name="reminder" max="20" value="0"/>
					</div>
					<input type="submit" value="Add"/>
				</form>
				<div id="reminder" style="margin-top: 10px"></div>
			</div>
			<script src="${scriptSrc}"></script>
		</body>
	</html>
	`
	res.send(response)
})

wss.on('connection', (ws, upgradeReq) => {
	const req = {	//Fake a request that we would normally see over HTTP
		originalUrl: '/',
		pathname: '/',
		headers: upgradeReq.headers
	}
	sessionHandler(req, {}, () => {	//Process session once a client connects
		ws.on('message', (msg) => {
			console.log(msg)
			const parsed = url.parse(msg)
			routeMessage(ws, parsed, req.session)
		})
	})
})

//Simple router
const routeMessage = (ws, request, session) => {
	const query = qs.parse(request.query)
	//The example app only has these functions
	const outputCallback = (out) => {
		ws.send(JSON.stringify(out))
	}

	if(request.pathname === '/add'){
		addRoute(query, session, outputCallback)
	}
	else if(request.pathname === '/delete'){
		deleteRoute(query, session, outputCallback)
	}
	else if(request.pathname === '/intro'){
		introRoute(query, session, outputCallback)
	}
}

//Display intro message
const introRoute = (query, session, cb) => {
	setTimeout(() => {	// Add a delay to highlight the content being loaded without user interaction
		cb([{
			action: 'replace',
			container: 'intro',
			content: '<h3>Just use the form below to add a todo<h3>'
		}])
	}, 2000)
}

//Deleting a todo
const deleteRoute = (query, session, cb) => {
	const filtered = session.todos.filter((item) => {
		return item.id !== parseInt(query.id, 10)
	})
	session.todos = filtered
	session.save(() => {
		const response = [{
			action: 'replace',
			container: 'lists',
			content: renderTodos(session)
		},{
			action: 'replace',
			container: 'count',
			content: renderTodoCount(session)
		}]
		cb(response)
	})
}

//Adding a todo
const addRoute = (query, session, cb) => {
	if(!session.todos){
		session.todos = []
	}
	const todo = {
		id: (new Date).getTime(),
		description: query.description,
		reminder: query.reminder
	}
	session.todos.push(todo)
	session.save(() => {
		const response = [{
			action: 'append',
			container: 'lists',
			content: '<li class="list-group-item">' + todo.description + ' <button class="pull-right" data-url="/delete?id=' + todo.id + '">Delete</button></li>'
		},{
			action: 'replace',
			container: 'count',
			content: renderTodoCount(session)
		}]
		if(todo.reminder !== '0'){
			console.log('scheduling reminder')
			sendReminder(todo, cb)
		}
		cb(response)
	})
}

//Send a message back to the client at some point in the future.
//cb() is the send function on the websocket so we can call it as many times as we need.
const sendReminder = (todo, cb) => {
	setTimeout(() => {
		const reminderResponse = [{
			action: 'replace',
			container: 'reminder',
			content: '<div class="alert alert-info" role="alert" onclick="this.style.display = \'none\';">Reminder for ' + todo.description + '!</div>'
		}]
		cb(reminderResponse)
	}, todo.reminder * 1000)
}

//Generate the HTML for all the todos
const renderTodos = (sess) => {
	if(!sess.todos){
		return ''
	}
	return sess.todos.reduce((acc, todo) => {
		acc += '<li class="list-group-item">' + todo.description + ' <button class="pull-right" data-url="/delete?id=' + todo.id + '">Delete</button></li>'
		return acc
	}, '')
}

//Generate the HTML for the number of todos
const renderTodoCount = (sess) => {
	const number = sess.todos ? sess.todos.length : 0
	return `You have <strong>${number}</strong> todos right now.`
}
